[ID-808] [FEATURE] Core BB auth support

jsafoan opened the pull request:
## Description
This PR adds updates the auth middleware to support the new Rokwire ID tokens, generated by the [Core BB](https://github.com/rokwire/core-building-block). Based on the issuer, the auth middleware will parse and validate the Core BB tokens, while still maintaining backwards compatibility for the Shibboleth and Phone tokens. 
An additional security scheme ```CoreTokenAuth``` has been added to all the service endpoints with ```ApiKeyAuth``` to validate anonymous tokens as well.

In addition, some of the existing code for validation has been refactored into helper functions for code reuse.
The additional environment variables that need to be added are included in [README.md](https://github.com/rokmetro/rokwire-building-blocks-api-fork/pull/5/files#diff-bdb0236f03e37fd81e52a79d67d8f2592244782b9d94455b69d09b1e735c10ca)

Fixes #808 

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [ ] Within a week
- [x] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [x] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My change requires updating the documentation.
- [x] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [x] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

ywkim312 commented:
I got following error while testing it. Could you please check?

token_info = func(request, required_scopes)
  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/security.py", line 267, in wrapper
    token_info = apikey_info_func(apikey, required_scopes=required_scopes)
  File "/usr/local/lib/python3.9/site-packages/auth_middleware/__init__.py", line 199, in verify_apikey
    keys = os.getenv('ROKWIRE_API_KEY').strip().split(',')
AttributeError: 'NoneType' object has no attribute 'strip'

shurwit commented:
> How do I test CoreAuthToken, how to generate it?

Hi @ywkim312, thanks for the review! For the core auth token, currently the easiest method of getting a token is to use the phone authentication. Please try the following:

1. Go to https://api-dev.rokwire.illinois.edu/core/doc/ui/index.html
2. Select the "Development server" from the "Servers" dropdown
3. Expand the `POST ​/services​/auth​/login` endpoint
4. Select "Try it out"
5. Send the following request body with your own phone number in the "phone" field in "creds"
```json
{
  "auth_type": "twilio_phone",
  "app_type_identifier": "edu.illinois.rokwire",
  "org_id": "0a2eff20-e2cd-11eb-af68-60f81db5ecc0",
  "creds": {
    "phone": "+12223334444"
  },
  "preferences": {
    "sample": "value"
  }
}
```
6. The response should be a message indicating that a verification code was sent, and you should receive a verification code on your phone via SMS
7. Send the following request body with your own phone number in the "phone" and the verification code you received in the "code" field in "creds"
```json
{
  "auth_type": "twilio_phone",
  "app_type_identifier": "edu.illinois.rokwire",
  "org_id": "0a2eff20-e2cd-11eb-af68-60f81db5ecc0",
  "creds": {
    "phone": "+12223334444",
    "code": "123456"
  }
}
```
8. The response you receive here should look like this where the "access_token" field is the token to test with:
```json
{
  ...
  "token": {
    "access_token": "<YOUR_ACCESS_TOKEN_HERE>",
    "refresh_token": "<YOUR_REFRESH_TOKEN_HERE>",
    "token_type": "Bearer"
  }
}
```
9. If your token expires, you can now repeat these steps to login again

Please let me know if you have any questions or run into any issues getting a token. 

Thanks!

ywkim312 commented:
This is my commend running for docker. Please let me know if there is something missing 
docker run --name profiles -e MONGO_PROFILE_URL=mongodb://xx.xx.xx.xx:27017 -e MONGO_PII_URL=mongodb://xx.xx.xx.xx:27017 -e ROKWIRE_AUTH_HOST=https://api-dev.rokwire.illinois.edu/core -e ROKWIRE_API_KEY=xxxxxxx-xxxxxx-xxxx -e ROKWIRE_AUTH_KEY_PATH=/tps/auth-keys -e ROKWIRE_API_CLIENTID=rokwire -p 5000:5000

jsafoan commented:
> I believe that the error happened before even get to the print statement. I will work on another project for the afternoon so I might not be able to response promptly. Could I ask you a favor? Could you test it by building a docker and let me know the run command when you test it to work?

Looks like in your Docker command you have the env variable as ```ROKWIRE_API_CLIENTID``` instead of ```ROKWIRE_API_CLIENT_ID```. That would lead to the invalid audience error you are seeing.

shurwit commented:
> I rebuilt the docker adding the print statement but it did not print and got this error 2021-10-11T17:19:15.669Z WARNING [Dummy-1 ] : auth_middleware - jwt error on decode. message = Invalid audience
> 
> I believe that the error happened before even get to the print statement. I will work on another project for the afternoon so I might not be able to response promptly. Could I ask you a favor? Could you test it by building a docker and let me know the run command when you test it to work?

Hi @ywkim312, I am trouble with my local Docker environment right now but I have confirmed that if you provide the following environment variables, including fixing the `ROKWIRE_API_CLIENT_ID` variable name as @jsafoan mentioned above it should work as expected. Please let us know if you are still having trouble after that change. Thanks!
```
MONGO_PROFILE_URL=mongodb://127.0.0.1:27017
MONGO_PII_URL=mongodb://127.0.0.1:27017
ROKWIRE_API_KEY=<API_KEY>
ROKWIRE_AUTH_HOST=https://api-dev.rokwire.illinois.edu/core
ROKWIRE_AUTH_KEY_PATH=/tps/auth-keys
ROKWIRE_API_CLIENT_ID=rokwire
```

shurwit commented:
> Thank you! It worked well. However, I only tested for profiles part. So we definately need Bing's help in event manger to be tested with this.

Great, thanks for the confirmation! We will wait for @bingzhang's review.

bingzhang commented:
> > > > auth
> > > 
> > > 
> > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > 
> > 
> > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> 
> Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.

I added +1 before the phone number. I still cannot receive the message.

> > > > auth
> > > 
> > > 
> > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > 
> > 
> > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> 
> Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.

I added +1 before the phone number, but I still cannot receive the code by the message. Do I change the `org_id`?

bingzhang commented:
> > > > > > auth
> > > > > 
> > > > > 
> > > > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > > > 
> > > > 
> > > > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> > > 
> > > 
> > > Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.
> > 
> > 
> > I added +1 before the phone number. I still cannot receive the message.
> > > > > > auth
> > > > > 
> > > > > 
> > > > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > > > 
> > > > 
> > > > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> > > 
> > > 
> > > Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.
> > 
> > 
> > I added +1 before the phone number, but I still cannot receive the code by the message. Do I change the `org_id`?
> 
> @bingzhang Are you getting the following success response when you make the phone login request?
> 
> ```
> {
>   "message": "verification code sent successfully",
>   "params": null
> }
> ```
> 
click the `Execute` button to send request? I did not receive any message.
> No you don't have to change anything else but the phone num
<img width="769" alt="Screen Shot 2021-10-13 at 5 06 30 PM" src="https://user-images.githubusercontent.com/5349154/137219140-c81c506e-368b-4394-a59b-bf3b4ba740e7.png">
ber in the request body

jsafoan commented:
> > > > > > > auth
> > > > > > 
> > > > > > 
> > > > > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > > > > 
> > > > > 
> > > > > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> > > > 
> > > > 
> > > > Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.
> > > 
> > > 
> > > I added +1 before the phone number. I still cannot receive the message.
> > > > > > > auth
> > > > > > 
> > > > > > 
> > > > > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > > > > 
> > > > > 
> > > > > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> > > > 
> > > > 
> > > > Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.
> > > 
> > > 
> > > I added +1 before the phone number, but I still cannot receive the code by the message. Do I change the `org_id`?
> > 
> > 
> > @bingzhang Are you getting the following success response when you make the phone login request?
> > ```
> > {
> >   "message": "verification code sent successfully",
> >   "params": null
> > }
> > ```
> 
> click the `Execute` button to send request? I did not receive any message.
> 
> > No you don't have to change anything else but the phone num
> 
> <img alt="Screen Shot 2021-10-13 at 5 06 30 PM" width="769" src="https://user-images.githubusercontent.com/5349154/137219140-c81c506e-368b-4394-a59b-bf3b4ba740e7.png">
> 
> ber in the request body

Once you hit execute, you can scroll down to see the response you got from the server. Also, please make sure you have selected the development server from the dropdown at the top of Swagger UI.

bingzhang commented:
> > > > > > > auth
> > > > > > 
> > > > > > 
> > > > > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > > > > 
> > > > > 
> > > > > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> > > > 
> > > > 
> > > > Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.
> > > 
> > > 
> > > I added +1 before the phone number. I still cannot receive the message.
> > > > > > > auth
> > > > > > 
> > > > > > 
> > > > > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > > > > 
> > > > > 
> > > > > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> > > > 
> > > > 
> > > > Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.
> > > 
> > > 
> > > I added +1 before the phone number, but I still cannot receive the code by the message. Do I change the `org_id`?
> > 
> > 
> > @bingzhang Are you getting the following success response when you make the phone login request?
> > ```
> > {
> >   "message": "verification code sent successfully",
> >   "params": null
> > }
> > ```
> 
> click the `Execute` button to send request? I did not receive any message.
> 
> > No you don't have to change anything else but the phone num
> 
> <img alt="Screen Shot 2021-10-13 at 5 06 30 PM" width="769" src="https://user-images.githubusercontent.com/5349154/137219140-c81c506e-368b-4394-a59b-bf3b4ba740e7.png">
> 
> ber in the request body

I also try the curl command. I have an error.


> > > > > > > > auth
> > > > > > > 
> > > > > > > 
> > > > > > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > > > > > 
> > > > > > 
> > > > > > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> > > > > 
> > > > > 
> > > > > Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.
> > > > 
> > > > 
> > > > I added +1 before the phone number. I still cannot receive the message.
> > > > > > > > auth
> > > > > > > 
> > > > > > > 
> > > > > > > Hi @bingzhang can you check that if you are still seeing the error after getting a new token following these [steps](https://github.com/rokwire/rokwire-building-blocks-api/pull/813#issuecomment-939119435) Just want to make sure the token isn't actually expired.
> > > > > > 
> > > > > > 
> > > > > > thanks. I followed the instructions but did not receive the smg. usually, how long it will take to receive the message?
> > > > > 
> > > > > 
> > > > > Can you make sure to add a +1 before your phone number and try again with the request body as mentioned in the instructions above? The code should be sent out almost immediately.
> > > > 
> > > > 
> > > > I added +1 before the phone number, but I still cannot receive the code by the message. Do I change the `org_id`?
> > > 
> > > 
> > > @bingzhang Are you getting the following success response when you make the phone login request?
> > > ```
> > > {
> > >   "message": "verification code sent successfully",
> > >   "params": null
> > > }
> > > ```
> > 
> > 
> > click the `Execute` button to send request? I did not receive any message.
> > > No you don't have to change anything else but the phone num
> > 
> > 
> > <img alt="Screen Shot 2021-10-13 at 5 06 30 PM" width="769" src="https://user-images.githubusercontent.com/5349154/137219140-c81c506e-368b-4394-a59b-bf3b4ba740e7.png">
> > ber in the request body
> 
> Once you hit execute, you can scroll down to see the response you got from the server. Also, please make sure you have selected the development server from the dropdown at the top of Swagger UI.

after changing to development, I received the code. But I have this error.

```
 curl -X POST "https://api-dev.rokwire.illinois.edu/core/services/auth/login" -H "accept: application/json" -H "Content-Type: application/json" -d "{\"auth_type\":\"twilio_phone\",\"app_type_identifier\":\"edu.illinois.rokwire\",\"org_id\":\"0a2eff20-e2cd-11eb-af68-60f81db5ecc0\",\"creds\":{\"phone\":\"+12176077050\",\"code\":\"365577\"},\"preferences\":{\"sample\":\"value\"}}"
500 - Error logging in: invalid response: status_code=404, error={"code": 20404, "message": "the requested resource /services/va4567d8fed395093ff838b14a34b0fe3c/verificationcheck was not found", "more_info": "https://www.twilio.com/docs/errors/20404", "status": 404}MacBook-Pro:~ bing$ 
```

bingzhang commented:
> > after changing to development, I received the code. But I have this error.
> > curl -X POST "https://api-dev.rokwire.illinois.edu/core/services/auth/login" -H "accept: application/json" -H "Content-Type: application/json" -d "{"auth_type":"twilio_phone","app_type_identifier":"edu.illinois.rokwire","org_id":"0a2eff20-e2cd-11eb-af68-60f81db5ecc0","creds":{"phone":"+12176077050","code":"365577"},"preferences":{"sample":"value"}}"
> > 500 - Error logging in: invalid response: status_code=404, error={"code": 20404, "message": "the requested resource /services/va4567d8fed395093ff838b14a34b0fe3c/verificationcheck was not found", "more_info": "https://www.twilio.com/docs/errors/20404", "status": 404}MacBook-Pro:~ bing$
> 
> Hi @bingzhang, you will receive this error if you try to submit the code again after it has already been used. Please try starting over from step 5 above (make the request with only your phone number and no code), then try sending the request with the new code you received. Please let us know if you continue to have trouble after that. Thanks!

thanks. I just sent the request in step 5. But I did not receive the code again.

bingzhang commented:
> > {"auth_type":"twilio_phone","app_type_identifier":"edu.illinois.rokwire","org_id":"0a2eff20-e2cd-11eb-af68-60f81db5ecc0","creds":{"phone":"+12176077050","code":"941499"},"preferences":{"sample":"value"}}"
> 
> Hi @bingzhang, you should not need to run the curl command from the shell since it should be the same as what is being done in the web app. You should be able to see the response to your request below the "execute" button once you have clicked it.
> 
> Just to be sure, you are sending two requests. The first should include only your phone number in the creds. You should then receive a text with the verification code. Then you send the second request with both the phone number and the verification code in the creds. That second request should return the token you need. If you try to send that second request again, you should receive an error like the one you mentioned so please be sure you are only sending the code once. For example, if you hit "execute" on the web app and copy the curl command, then execute the curl command in the shell, the request would have been sent twice. Please try to find the response on the web page instead of using curl locally.
> 
> Let me know if you are still having trouble. Thanks!

I tested the postman to events building block. there is an unauthorized error for the post method. And I tested the events manager, the get tags failed.

shurwit commented:
> I tested the postman to events building block. there is an unauthorized error for the post method. And I tested the events manager, the get tags failed.

Hi @bingzhang, could you provide the error information for the get tags request? The unauthorized post is expected as we discussed.

shurwit commented:
> Sorry, please ignore the tag get failure. It uses the same API key. but some other configuration is messed up.
> 
> There is a post failure from events manager. It uses the `id_token` from user login, e.g., ` headers = { 'Content-Type': 'application/json', 'Authorization': 'Bearer ' + session["id_token"] }`
> 
> but it failed with `'{ "detail": "Invalid token", "status": 401, "title": "Unauthorized", "type": "about:blank" } '`

Hi @bingzhang, thanks for the update. Is the events service configured with the correct environment variables to validate the Shibboleth token being sent from the events manager? Please verify that you have the correct `SHIBBOLETH_HOST` and `SHIBBOLETH_CLIENT_ID` environment variables are set as follows:

```
SHIBBOLETH_HOST=shibboleth.illinois.edu
SHIBBOLETH_CLIENT_ID=<CLIENT ID OF SHIBBOLETH TOKEN TO BE VALIDATED>
```

ywkim312 commented:
> > Hi @sandeep-ps, it looks like we need an approval from you before we will be allowed to merge this PR. Please let us know if there is anyone else you would like to perform a review before this. If not, then please approve this when you get a chance. Thanks!
> 
> Sure. For App Config and Covid App Config changes, we will need @minump to review. I have added Minu.
> 
> Hi, Minu, Could you please review those changes related to App Config and Covid App Config when you get a chance? There are many details in this PR that'll help you to test/review. Please contact Yongwook, Bing, or Stephen if you have questions. Thanks.
> 
> @ywkim312, have you reviewed the Contributions BB and Logging BB changes too? I think at one place you have mentioned that you only tested Profile BB changes, and hence this question.
> 
> Hi, Yongwook, @bingzhang, and Minu,
> 
> I think we mainly need to make sure that all current endpoints work with both the Core BB token and the current Shibooleth ID token and API keys. I suppose that those have been verified. Thanks.

Hmm, weird, since profiles work, contributins should work as well since they are sharing the same auth logic. However, when I tried POST new contribution using Core Access Token I got the following error

"detail": "Provided token is not valid",
"status": 401,
"title": "Unauthorized",
"type": "about:blank"

sandeep-ps commented on appconfigservice/appconfig.yaml:
```diff
@@ -91,6 +91,7 @@ paths:
         Auth: Requires a valid API Key for access.
       security:
         - ApiKeyAuth: []
+        - CoreTokenAuth: []
```
Does this not provide easier access to all app configurations, especially as generating a Core access token is relatively straightforward when compared with the API key?

shurwit replied:
Hi @sandeep-ps, this is actually a very good point. We will have to reconsider how API keys are being used and validated on the Core BB side. Thanks for catching this!

shurwit replied:
Hi @sandeep-ps, I opened this issue https://github.com/rokwire/core-building-block/issues/244 to handle this issue on the Core BB. As you can see in the thread and related PR, we decided to validate the API key before allowing the user to generate a token at all. Once these Core BB changes have been approved and merged, I believe this issue will be resolved without needing any changes in this PR. This is because we can guarantee that anyone who has a valid token had a valid API key. Let me know if you see any issues with this approach. Thanks!

sandeep-ps commented on contributions/contribution.yaml:
```diff
@@ -782,4 +782,4 @@ components:
           scopes: {}
       x-tokenInfoUrl: https://api.github.com/user
-      description: The client must send a unexpired GitHub OAuth2 access_token in the Authorization header when making requests to API endpoints that require GitHub authentication.
+      description: The client must send a unexpired GitHub OAuth2 access_token in the Authorization header when making requests to API endpoints that require GitHub authentication.
```
Minor suggestion - I think this format change is not.

sandeep-ps commented on eventservice/events.yaml:
```diff
@@ -25,7 +25,7 @@ paths:
       security:
         - ApiKeyAuth: []
-        - UserAuth: []
+        - UserAndCoreTokenAuth: []
```
The previous UserAuth verifies users' access to an event using their Rokwire group membership. Users can view only public events and private group events that they have access to due to Rokwire Groups BB group memberships. Will that be affected by this new change? Same question for all other Events BB endpoints that replaces UserAuth with UserAndCoreTokenAuth.

jsafoan replied:
Currently, the groups BB only supports those checks for shibboleth tokens. We've added some changes [here](https://github.com/rokwire/rokwire-building-blocks-api/blob/ac52faff857dc3d52d2ec4e45e3bef4d62f2c1a1/eventservice/api/utils/group_auth.py#L40-L44) to extend those checks to the core tokens as well. But additional changes need to be made on the groups BB as well.

sandeep-ps commented:
> We have added a wrapper to apply two different security auths(UserAuth and CoreTokenAuth) with the same Auth header. You can test this change mainly in the event service.
However, the GithubAuth is a little more complicated, as that is being handled externally by OAuth endpoints. We would have to migrate all of that into auth middleware, if we want to support multiple token auths for contributing service. We're debating if that is worthwhile, as we're not even sure where the service being used right now. Do you have any idea about that?

Let me try to add some details here. Contributions BB is currently deployed in the DEV instance will be used to accept/review external contributions like third-party services/integrations. It's using GitHub Auth since we believe this is mostly meant for other developers who want to contribute their services/UI components to Rokwire. GitHub apps are using OAuth2 and not OIDC and hence we are using this approach. I don't think we will have any issues if this can continue to use GitHub directly.

CC: @ywkim312

ywkim312 commented:
> > @shurwit the token creation doesn't work anymore for me saying that the device information is missing.
> > 400 - Error validating request: request body has an error: doesn't match the schema: Error at "/device": property "device" is missing
> 
> Hi @ywkim312, some changes were deployed last night. I have updated my example above with a new example request. Please let me know if you are still having trouble. Thanks!

Still having the same error at the very first step.

ywkim312 commented:
> > @shurwit Could I ask a question? When I generate the phone based access token for profiles not core token, it doesn't give me the access token anymore but only gives me id token. Is it possible for you to check it out? Or should I ask someone else?
> 
> Hi @ywkim312, yes it is just a slight difference in the naming conventions between the old auth service and the Core BB, however they essentially represent the same thing. The auth service in this repository and it's phone tokens will be deprecated in ROKWIRE 3.0 and the Core BB tokens will replace them. Due to this it may not make sense to worry about these differences too much. Let me know if you have any other questions or if I missed your point. Thanks!

I see. Thank you for your answer. If so, it makes an error in pii post in profiles building block when I used the core access token. Here's the message

[2021-10-19 20:34:18 +0000] [9] [ERROR] Error handling request /profiles/pii
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/gunicorn/workers/base_async.py", line 55, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/usr/local/lib/python3.9/site-packages/gunicorn/workers/ggevent.py", line 143, in handle_request
    super().handle_request(listener_name, req, sock, addr)
  File "/usr/local/lib/python3.9/site-packages/gunicorn/workers/base_async.py", line 106, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/local/lib/python3.9/site-packages/connexion/apps/abstract.py", line 248, in __call__
    return self.app(environ, start_response)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2463, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2449, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1866, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/decorator.py", line 48, in wrapper
    response = function(request)
  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/security.py", line 299, in wrapper
    return function(request)
  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/uri_parsing.py", line 173, in wrapper
    response = function(request)
  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/validation.py", line 164, in wrapper
    error = self.validate_schema(data, request.url)
  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/validation.py", line 205, in validate_schema
    self.validator.validate(data)
  File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 248, in validate
    for error in self.iter_errors(*args, **kwargs):
  File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 224, in iter_errors
    for error in errors:
  File "/usr/local/lib/python3.9/site-packages/jsonschema/_validators.py", line 329, in properties
    yield from validator.descend(
  File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 240, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 224, in iter_errors
    for error in errors:
  File "/usr/local/lib/python3.9/site-packages/jsonschema/_validators.py", line 391, in oneOf
    more_valid = [
  File "/usr/local/lib/python3.9/site-packages/jsonschema/_validators.py", line 393, in <listcomp>
    if validator.evolve(schema=each).is_valid(instance)
  File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 270, in is_valid
    error = next(self.iter_errors(instance), None)
  File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 224, in iter_errors
    for error in errors:
  File "/usr/local/lib/python3.9/site-packages/connexion/json_schema.py", line 60, in validate_type
    yield ValidationError(_utils.types_msg(instance, types))
AttributeError: module 'jsonschema._utils' has no attribute 'types_msg'

ywkim312 commented:
> > > > @shurwit Could I ask a question? When I generate the phone based access token for profiles not core token, it doesn't give me the access token anymore but only gives me id token. Is it possible for you to check it out? Or should I ask someone else?
> > > 
> > > 
> > > Hi @ywkim312, yes it is just a slight difference in the naming conventions between the old auth service and the Core BB, however they essentially represent the same thing. The auth service in this repository and it's phone tokens will be deprecated in ROKWIRE 3.0 and the Core BB tokens will replace them. Due to this it may not make sense to worry about these differences too much. Let me know if you have any other questions or if I missed your point. Thanks!
> > 
> > 
> > I see. Thank you for your answer. If so, it makes an error in pii post in profiles building block when I used the core access token. Here's the message
> > [2021-10-19 20:34:18 +0000] [9] [ERROR] Error handling request /profiles/pii Traceback (most recent call last): File "/usr/local/lib/python3.9/site-packages/gunicorn/workers/base_async.py", line 55, in handle self.handle_request(listener_name, req, client, addr) File "/usr/local/lib/python3.9/site-packages/gunicorn/workers/ggevent.py", line 143, in handle_request super().handle_request(listener_name, req, sock, addr) File "/usr/local/lib/python3.9/site-packages/gunicorn/workers/base_async.py", line 106, in handle_request respiter = self.wsgi(environ, resp.start_response) File "/usr/local/lib/python3.9/site-packages/connexion/apps/abstract.py", line 248, in **call** return self.app(environ, start_response) File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2463, in **call** return self.wsgi_app(environ, start_response) File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2449, in wsgi_app response = self.handle_exception(e) File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1866, in handle_exception reraise(exc_type, exc_value, tb) File "/usr/local/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise raise value File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2446, in wsgi_app response = self.full_dispatch_request() File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1951, in full_dispatch_request rv = self.handle_user_exception(e) File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1820, in handle_user_exception reraise(exc_type, exc_value, tb) File "/usr/local/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise raise value File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1949, in full_dispatch_request rv = self.dispatch_request() File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1935, in dispatch_request return self.view_functions[rule.endpoint](**req.view_args) File "/usr/local/lib/python3.9/site-packages/connexion/decorators/decorator.py", line 48, in wrapper response = function(request) File "/usr/local/lib/python3.9/site-packages/connexion/decorators/security.py", line 299, in wrapper return function(request) File "/usr/local/lib/python3.9/site-packages/connexion/decorators/uri_parsing.py", line 173, in wrapper response = function(request) File "/usr/local/lib/python3.9/site-packages/connexion/decorators/validation.py", line 164, in wrapper error = self.validate_schema(data, request.url) File "/usr/local/lib/python3.9/site-packages/connexion/decorators/validation.py", line 205, in validate_schema self.validator.validate(data) File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 248, in validate for error in self.iter_errors(*args, **kwargs): File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 224, in iter_errors for error in errors: File "/usr/local/lib/python3.9/site-packages/jsonschema/_validators.py", line 329, in properties yield from validator.descend( File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 240, in descend for error in self.evolve(schema=schema).iter_errors(instance): File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 224, in iter_errors for error in errors: File "/usr/local/lib/python3.9/site-packages/jsonschema/_validators.py", line 391, in oneOf more_valid = [ File "/usr/local/lib/python3.9/site-packages/jsonschema/_validators.py", line 393, in if validator.evolve(schema=each).is_valid(instance) File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 270, in is_valid error = next(self.iter_errors(instance), None) File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 224, in iter_errors for error in errors: File "/usr/local/lib/python3.9/site-packages/connexion/json_schema.py", line 60, in validate_type yield ValidationError(_utils.types_msg(instance, types)) AttributeError: module 'jsonschema._utils' has no attribute 'types_msg'
> 
> Hi @ywkim312, you can skip testing the core tokens on profile service, because it will be deprecated and soon be replaced by the core building block. I will undo any of the changes made to the security schemes for the profile endpoints and push the update.

I see, if so, I wouldn't worry but want @sandeep-ps 's confirmation on this as well.

shurwit commented on eventservice/api/utils/group_auth.py:
```diff
@@ -1,18 +1,19 @@
-    # user UserAuth request
-    if 'user_token_data' in g:
+    if 'user_token' in g:
         include_private_events = True
-        auth_resp = g.user_token_data
-        uin = auth_resp.get('uiucedu_uin')
-        url = "%s%s/groups" % (cfg.GROUPS_BUILDING_BLOCK_ENDPOINT, uin)
+        id_token = g.user_token
+        url = cfg.GROUPS_BUILDING_BLOCK_HOST + "/api/user/group-memberships"
         headers = {"Content-Type": "application/json",
-                  "ROKWIRE_GS_API_KEY": cfg.ROKWIRE_GROUPS_API_KEY}
+                    "Authorization": "Bearer " + id_token}
```
Hi @sandeep-ps @bingzhang, I wanted to give a bit more context for these changes here. The issue is that we want to be able to support auth types other than Shibboleth for private group events. The Groups BB already knows how to parse all possible login types for the Core tokens and also accepts and understands how to handle Shibboleth tokens. Rather than trying to duplicate that complicated work, we thought it would be best to simply forward the user tokens to a new Groups BB endpoint protected by user token auth that parses the user ID from the token and returns the same data as the old endpoint. Please let us know if either of you have any questions or concerns about this approach. Thanks!

edthedev commented on lib/auth-middleware/auth_middleware/__init__.py:
```diff
@@ -245,31 +284,146 @@ def verify_userauth(id_token, group_name=None, internal_token_only=False):
+
 
 def use_security_token_auth(func):
     func._use_security_token_auth = True
```
This deserves a comment answering why this is always set to True.

edthedev replied:
Looks like a decorator - maybe a DocString mentioning when/where it should be used.

sandeep-ps commented on lib/auth-middleware/auth_middleware/__init__.py:
```diff
@@ -208,34 +252,30 @@ def verify_userauth(id_token, group_name=None, internal_token_only=False):
         keyset = None
         target_client_ids = None
 
-        if issuer == ROKWIRE_ISSUER:
```
Hi @shurwit and @jsafoan, 

It looks like we missed this particular case when reviewing this PR. The Events Manager uses the Rokwire issuer and the tokens signed by it to talk to the Events Building Block without user login. So, at present, these are like service tokens used for nightly upload/update of events. Since this particular "if case" was removed in the PR, currently uploading WebTools events to Events Manager doesn't work in the DEV instance, affecting backward compatibility. Our team could fix it by bringing this back, but I wanted to get your thoughts before we do that. Thanks.

shurwit replied:
Hi @sandeep-ps, sorry we missed that! Yes you are welcome to add this back in for now. At some point we will use the service accounts from the Core BB that we are discussing here https://github.com/rokwire/core-building-block/issues/306 to replace this, but for now it should be fine to continue supporting this mechanism. Feel free to make any changes needed to get this working again for now, and we can reconcile any issues when we discuss migrating this functionality to Core service accounts.

sandeep-ps replied:
No issues. We also missed it during the review. Will get this added.

sandeep-ps commented on eventservice/api/utils/group_auth.py:
```diff
@@ -26,14 +26,12 @@ def get_group_ids():
     include_private_events = False
-    # user UserAuth request
-    if 'user_token_data' in g:
+    if 'user_token' in g and not g.user_token_data.get('anonymous'):
```
Hi, @shurwit, 

The new group membership endpoint is returning 401 with the tokens signed by Rokwire issuer api.rokwire.illinois.edu. As you know, this was the previous endpoint - /gr/api/int/user/{UIN}/groups. When UIN is None (since there is no logged-in user here), the previous endpoint would return an empty list along with success code 200. Should we continue to use that endpoint while the service tokens feature is being added? Or do you have other suggestions? This is also applicable to get_group_ids method above. Please let us know. Thanks.

CC: @ywkim312

shurwit replied:
Hi @sandeep-ps @ywkim312, yes you are correct that the new endpoint will only accept core tokens. I think our best bet for now is to determine if the token is a Core token here. If it is, we will use the new endpoint as it does now. If not, it will use the old endpoint. To do this we could check the issuer of the token. Let me know if you see any issues with this approach or if you have any questions. Thanks!

sandeep-ps replied:
Okay. Thanks. Yes, it makes sense to check for the issuer and make different calls based on that.

@ywkim312, let's have a chat to get this clarified.

shurwit replied:
Sorry @sandeep-ps, I realized I made a mistake in my last post. The new endpoint will accept either Core or Shibboleth tokens. It is important that we continue sending Core tokens to this endpoint to support non-Shibboleth authenticated users. I do not think it really matters which endpoint we use for Shibboleth tokens, but I would suggest leaving them on the new endpoint just to be safe. The Rokwire issuer tokens will need to go to the old endpoint as I described. Hopefully this helps!

sandeep-ps replied:
Yes, this really helps. Thank you!
